### PR TITLE
Infrastructure for Camel Kamelets -> Assemblies

### LINKED ISSUE #387: camel: Add infrastructure for constructing assemblies

### PR DESCRIPTION
**Pull request by fhcarter**
Fixes #387 

This adds infrastructure in the form of a specialized Gradle task to convert the published Apache Camel kamelets into Vantiq assemblies.

At this point, we are adding the `kamelet2assemblies` subproject and the task infrastructure to begin the interpretation. We are not yet building assemblies;  instead, we convert the kamelet definition obtained from the maven repo (via a gradle dependency) to some proof of concept artifacts.  The next step will be to begin to convert those artifacts into Vantiq projects/assemblies.

Since this will be some ongoing work, I have created a _staging_ branch, `fhc-vantiqAssemblies`, into which we'll merge the various parts until the whole is ready for release. This _staging_ branch is the branch into which this PR will be merged.  Once complete, we can merge the staging branch into master.  But there's no real point on inflicting a bunch of partially completed stuff that doesn't work on the world.

For background, a _kamelet_ is a packaging of some standard routes for accessing various (about 80 number) different data sources and sinks.  They do this in a parameterized manner, and are packaged for use in a Camel-K environment (a set of operators deployed into Kubernetes that allows the system to sort of construct itself).  

The Apache Camel connector will be deployable via these assemblies, allowing the relatively easy deployment of a means to attach these 80 runtime environments to Vantiq.  Each of the 80 data environments has a _source_ and a _sink_ variant, so we end up with 160 different connection possibilities (80 environments * 2 _directions_).

As noted, this is the beginning.  Subsequent PR's will push the various artifacts closer to project/assembly definitions.

In this PR, we add the subproject and create a `gradle` task (type) that will perform the conversion based on the dependencies describe in the invoking project.   This task will product a bunch of build artifacts that will eventually become the sources for the assemblies (via subsequent PRs).

### CODE REVIEW

File: buildSrc/build.gradle [hunk 1]
```diff
@@ -0,0 +1,34 @@
+
```
Comment (fhcarter) [inline buildSrc/build.gradle:1]: Construct build.gradle to for buildSrc to build the task we need.

File: buildSrc/src/main/java/io/vantiq/extsrc/assy/tasks/AssemblyGen.java [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+    public static final String YML_FILE_TYPE = ".yml";
+    public static final String SOURCE_KAMELET_PREFIX = "-source.kamelet";
+    public static final String SINK_KAMELET_PREFIX = "-sink.kamelet";
+    public static final String SOURCE_KAMELET_YAML_SUFFIX = SOURCE_KAMELET_PREFIX + YAML_FILE_TYPE;
+    public static final String SOURCE_KAMELET_YML_SUFFIX = SOURCE_KAMELET_PREFIX + YML_FILE_TYPE;
+    public static final String SINK_KAMELET_YAML_SUFFIX = SINK_KAMELET_PREFIX + YAML_FILE_TYPE;
+    public static final String SINK_KAMELET_YML_SUFFIX = SINK_KAMELET_PREFIX + YML_FILE_TYPE;
+    
+    public final Project project;
+
```
Comment (fhcarter) [inline buildSrc/src/main/java/io/vantiq/extsrc/assy/tasks/AssemblyGen.java:58]: Gradle stuff

File: buildSrc/src/main/java/io/vantiq/extsrc/assy/tasks/AssemblyGen.java [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+        // From the gradle data, construct the list of resolved dependencies & build a class loader to use
+        Configuration conf = project.getConfigurations().getByName("runtimeClasspath");
+        Set<File> files = conf.resolve();
+        log.info("File count from runtimeClasspath: {}", files.size());
+        File[] jars = files.toArray(new File[0]);
+        ArrayList<URL> urls = new ArrayList<>(jars.length);
+        for (File f: jars) {
+            log.debug("Adding {} to list of URLs", f.toURI().toURL());
+            urls.add(f.toURI().toURL());
+        }
```
Comment (fhcarter) [inline buildSrc/src/main/java/io/vantiq/extsrc/assy/tasks/AssemblyGen.java:108]: The class loader below is constructed to load the dependencies as declared in the gradle file.  We use this to obtain the kamelet definitions.

File: buildSrc/src/main/java/io/vantiq/extsrc/assy/tasks/AssemblyGen.java [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+                        DumpSettings dumpSettings = DumpSettings.builder()
+                                                                .setIndent(4)
+                                                                // FLOW (default)serializes any embedded maps.
+                                                                // This "recurses".
+                                                                .setDefaultFlowStyle(FlowStyle.BLOCK)
+                                                                .build();
+                        Dump dumper = new Dump(dumpSettings);
+                        
+                        // Now, extract the route template.  This will become our route document after we change
+                        // kamelet://source or kamelet::sink to vantiq server URLs
```
Comment (jcguion) [inline buildSrc/src/main/java/io/vantiq/extsrc/assy/tasks/AssemblyGen.java:174]: nit - should it be rather kamelet:source and kamelet:sink ?

File: buildSrc/src/main/java/io/vantiq/extsrc/assy/tasks/AssemblyGen.java [hunk 5]
```diff
@@ -0,0 +1,336 @@
...
+                        if (!Files.exists(thisKameletDirectory)) {
+                            log.info("Creating kamelet-specific project directory: {}", thisKameletDirectory);
+                            thisKameletDirectory = Files.createDirectories(thisKameletDirectory);
+                        }
+                        log.info("\nResults for Kamelet: {}", kamName);
+                        log.info("\n>>>   Properties:");
+                        props.forEach((name, val) -> {
+                            //noinspection unchecked
+                            log.info("    Name: {} :: [title: {}, type: {}, desc: {}]", name,
+                                     ((Map<String, String>) val).get("t!itle"),
```
Comment (jcguion) [inline buildSrc/src/main/java/io/vantiq/extsrc/assy/tasks/AssemblyGen.java:195]: typo `title`

**Review by jcguion — APPROVED** [review]
LGTM

**Pull request merged**